### ISSUE TITLE
Server filters

### ISSUE THREAD

**Issue by max-ieremenko** [max-ieremenko/ServiceModel.Grpc#53]
The concept is simple: server filter is a hook for service method invocation. It can work together with gRPC server interceptors, but it is not interceptor.

``` c#
internal sealed class LoggingServerFilter : IServerFilter
{
    private readonly ILoggerFactory _loggerFactory;

    public LoggingServerFilter(ILoggerFactory loggerFactory)
    {
        _loggerFactory = loggerFactory;
    }

    public async ValueTask InvokeAsync(IServerFilterContext context, Func<ValueTask> next)
    {
        // create logger with a service name
        var logger = _loggerFactory.CreateLogger(context.ServiceInstance.GetType().Name);

        // before service method invocation log input
        logger.LogInformation("Begin {0}", context.ContractMethodInfo.Name);
        foreach (var entry in context.Request)
        {
            logger.LogInformation("Input {0} = {1}", entry.Key, entry.Value);
        }

        try
        {
            // invoke all other filters in the stack and the service method
            await next().ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            // log exception
            logger.LogError("Method {0} failed: {1}", context.ContractMethodInfo.Name, ex);
            throw;
        }

        // after service method is finished log output
        logger.LogInformation("End {0}", context.ContractMethodInfo.Name);
        foreach (var entry in context.Response)
        {
            logger.LogInformation("Output {0} = {1}", entry.Key, entry.Value);
        }
    }
}
```

**How to register the filter.**

_Variant 1_: register as global one for all services:

``` c#
public void ConfigureServices(IServiceCollection services)
{
    // instance is transient
    services.AddTransient<LoggingServerFilter>();

    services.AddServiceModelGrpc(options =>
    {
        // register LoggingServerFilter
        options.Filters.Add(1, provider => provider.GetRequiredService<LoggingServerFilter>());
    });
}
```

_Variant 2_: register for a specific service only:

``` c#
public void ConfigureServices(IServiceCollection services)
{
    // instance is transient
    services.AddTransient<LoggingServerFilter>();

    services.AddServiceModelGrpcServiceOptions<Calculator>(options =>
    {
        // register LoggingServerFilter
        options.Filters.Add(1, provider => provider.GetRequiredService<LoggingServerFilter>());
    });
}
```

_Variant 3_: mark service with attribute:

``` c#
[ServerFilter(1, typeof(LoggingServerFilter))]
public sealed class Calculator : ICalculator
{
    public ValueTask<int> SumAsync(int x, int y, CancellationToken token)
    {
        return new ValueTask<int>(x + y);
    }
}
```

_Variant 4_: mark method with attribute:

``` c#
public sealed class Calculator : ICalculator
{
    [ServerFilter(1, typeof(LoggingServerFilter))]
    public ValueTask<int> SumAsync(int x, int y, CancellationToken token)
    {
        return new ValueTask<int>(x + y);
    }
}
```

_At runtime_ for SumAsync(1, 2) the log output will be

``` text
INFO Begin SumAsync
INFO Input x = 1
INFO Input y = 2
INFO End SumAsync
INFO Output result = 3
```

**The filter has complete control over the method call.**

``` c#
[ServerFilter(100, typeof(SumAsyncServerFilter))]
public ValueTask<int> SumAsync(int x, int y, CancellationToken token)
{
    // the filter must handle the call
    throw new NotImplementedException();
}

internal sealed class SumAsyncServerFilter : IServerFilter
{
    public ValueTask InvokeAsync(IServerFilterContext context, Func<ValueTask> next)
    {
        // var x = (int)context.Request[0];
        var x = (int)context.Request["x"];

        // var y = (int)context.Request[1];
        var y = (int)context.Request["y"];

        // do not invoke the real SumAsync
        // await next().ConfigureAwait(false);

        // context.Response[0] = x + y;
        context.Response["result"] = x + y;
        
        return new ValueTask(Task.CompletedTask);
    }
}

// at runtime SumAsync(1 + 2) will be 3
```

**Comment by max-ieremenko** [comment]
link #48: "Can Use FilterAttribute?"
link #47: "Is't possible to suppress an exception?"

**Comment by max-ieremenko** [comment]
released in 1.4.0, see [docs](https://max-ieremenko.github.io/ServiceModel.Grpc/server-filters.html) and [example](https://github.com/max-ieremenko/ServiceModel.Grpc/tree/master/Examples/ServerFilters)
